components/browser/menu2/src/main/java/mozilla/components/browser/menu2/adapter/icons/MenuIconViewHolder.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.menu2.adapter.icons
+
+import android.view.LayoutInflater
+import android.view.View
+import androidx.annotation.LayoutRes
+import androidx.constraintlayout.widget.ConstraintLayout
+import androidx.constraintlayout.widget.ConstraintSet
+import androidx.constraintlayout.widget.ConstraintSet.END
+import androidx.constraintlayout.widget.ConstraintSet.PARENT_ID
+import androidx.constraintlayout.widget.ConstraintSet.START
+import mozilla.components.browser.menu2.R
+import mozilla.components.browser.menu2.candidate.MenuIcon
+
+/**
+ * View holder with a [bind] method that passes the previously bound value.
+ */
+internal abstract class MenuIconViewHolder<T : MenuIcon>(
+    protected val parent: ConstraintLayout,
+    protected val inflater: LayoutInflater
+) {
+
+    @Suppress("Unchecked_Cast")
+    fun bindAndCast(newIcon: MenuIcon, oldIcon: MenuIcon?) {
+        bind(newIcon as T, oldIcon as T?)
```

Comment:
That's one of the unfortunate side-effects of using sealed classes. I don't have any simple solutions here besides trying to make `MenuIcon` an interface, but that might pose larger changes in how you match the view holders.